ros/src/actuation/vehicles/packages/as/nodes/ssc_interface/ssc_interface.cpp

```diff
@@ -0,0 +1,286 @@
+  ssc_feedbacks_sync_->registerCallback(
+      boost::bind(&SSCInterface::callbackFromSSCFeedbacks, this, _1, _2, _3, _4, _5, _6, _7));
+
+  // publishers to autoware
+  vehicle_status_pub_ = nh_.advertise<autoware_msgs::VehicleStatus>("vehicle_status", 10);
+  current_twist_pub_ = nh_.advertise<geometry_msgs::TwistStamped>("vehicle/twist", 10);
+
+  // publishers to SSC
+  steer_mode_pub_ = nh_.advertise<automotive_platform_msgs::SteerMode>("as/arbitrated_steering_commands", 10);
+  speed_mode_pub_ = nh_.advertise<automotive_platform_msgs::SpeedMode>("as/arbitrated_speed_commands", 10);
+  turn_signal_pub_ = nh_.advertise<automotive_platform_msgs::TurnSignalCommand>("as/turn_signal_command", 10);
+  gear_pub_ = nh_.advertise<automotive_platform_msgs::GearCommand>("as/gear_select", 1, true);
+}
+
+SSCInterface::~SSCInterface()
+{
+}
+
+void SSCInterface::run()
+{
+  while (ros::ok())
+  {
+    ros::spinOnce();
+    publishCommand();
+    rate_->sleep();
+  }
+}
+
+void SSCInterface::callbackFromVehicleCmd(const autoware_msgs::VehicleCmdConstPtr& msg)
+{
+  command_time_ = ros::Time::now();
+  vehicle_cmd_ = *msg;
+  command_initialized_ = true;
+}
+
+void SSCInterface::callbackFromEngage(const std_msgs::BoolConstPtr& msg)
+{
+  engage_ = msg->data;
+}
+
+void SSCInterface::callbackFromSSCModuleStates(const automotive_navigation_msgs::ModuleStateConstPtr& msg)
+{
+  if (msg->name.find("veh_controller") != std::string::npos)
+  {
+    module_states_ = *msg;  // *_veh_controller status is used for 'drive/steeringmode'
+  }
+}
+
+void SSCInterface::callbackFromSSCFeedbacks(const automotive_platform_msgs::VelocityAccelConstPtr& msg_velocity,
+                                            const automotive_platform_msgs::CurvatureFeedbackConstPtr& msg_curvature,
+                                            const automotive_platform_msgs::ThrottleFeedbackConstPtr& msg_throttle,
+                                            const automotive_platform_msgs::BrakeFeedbackConstPtr& msg_brake,
+                                            const automotive_platform_msgs::GearFeedbackConstPtr& msg_gear,
+                                            const pacmod_msgs::WheelSpeedRptConstPtr& msg_wheel_speed,
+                                            const pacmod_msgs::SystemRptFloatConstPtr& msg_steering_wheel)
+{
+  ros::Time stamp = msg_velocity->header.stamp;
+
+  // current speed
+  double speed =
+      !use_rear_wheel_speed_ ?
+          (msg_velocity->velocity) :
+          (msg_wheel_speed->rear_left_wheel_speed + msg_wheel_speed->rear_right_wheel_speed) * tire_radius_ / 2.;
+
+  // update adaptive gear ratio
+  adaptive_gear_ratio_ = agr_coef_a_ + agr_coef_b_ * speed * speed - agr_coef_c_ * msg_steering_wheel->output;
+  // current steering curvature
+  double curvature = !use_adaptive_gear_ratio_ ?
+                         (msg_curvature->curvature) :
+                         std::tan(msg_steering_wheel->output / adaptive_gear_ratio_) / wheel_base_;
```

Comment:
Will not be divided by 0?
Is there a guarantee that `adaptive_ger_ratio_` will not be 0?

Reply:
msg_steering_wheel->output could be 0 if the wheel is perfectly centered. This would cause a divide-by-zero error here, as @yn-mrse points out.

Reply:
@yn-mrse @JWhitleyAStuff @snehagn Thanks a lot. Fixed.
https://github.com/CPFL/Autoware/pull/1945/commits/ea6f8089fe671a23974833366bf74c73050ac10d